### PR TITLE
[RISCV] Add a rematerializable pseudo instruction for LUI+ADDI for global addresses.

### PR DESCRIPTION
**Pull request by topperc**
This allows register allocation to rematerialize these instead of spilling and reloading. We need to make it a single instruction due to limitations in rematerialization.
    
This pseudo is expanded to an LUI+ADDI pair between regalloc and post RA scheduling.

This improves the dynamic instruction count on 531.deepsjeng_r from spec2017 by 3.2% for the train dataset. 500.perlbench and 502.gcc see a 1% improvement. There are couple regressions, but they are 0.1% or smaller.

AArch64 has similar pseudo instructions like `MOVaddr`

Co-authored-by: Jesse Huang <jesse.huang@sifive.com>

Stacked on #93129

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVInstrInfo.td [hunk 1]
```diff
@@ -1311,6 +1311,22 @@ def : Pat<(FrameAddrRegImm (iPTR GPR:$rs1), simm12:$imm12),
 
 /// HI and ADD_LO address nodes.
 
+let Size = 8, isReMaterializable = 1 in
+def PseudoLIaddr : Pseudo<(outs GPR:$dst), (ins uimm20_lui:$hi, simm12:$lo), []>,
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.td:1315]: What about `PseudoLIAddr` or `PseudoLoadAddr`?

File: llvm/lib/Target/RISCV/RISCVInstrInfo.td [hunk 2]
```diff
@@ -1311,6 +1311,22 @@ def : Pat<(FrameAddrRegImm (iPTR GPR:$rs1), simm12:$imm12),
...
+                     Sched<[WriteIALU]>;
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.td]: Don't we need `SchedRead`s here?

**Comment by topperc** [comment]
@dtcxzyw can you do any measurements on this?

**Comment by dtcxzyw** [comment]
> @dtcxzyw can you do any measurements on this?

I am sorry I cannot help. It is weird to me that checksums of llvm-test-suite binaries don't change after this patch :(

File: llvm/lib/Target/RISCV/RISCVInstrInfo.td (hunk 2 above)
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.td]: There are no register operands so they wouldn't be associated with anything.

**Comment by topperc** [comment]
> > @dtcxzyw can you do any measurements on this?
> 
> I am sorry I cannot help. It is weird to me that checksums of llvm-test-suite binaries don't change after this patch :(

That is weird to me too.

File: llvm/lib/Target/RISCV/RISCVInstrInfo.td (hunk 2 above)
Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.td]: I don't know if we should model the second instruction `ADDI` here.

File: llvm/lib/Target/RISCV/RISCVInstrInfo.td (hunk 1 above)
Comment (preames) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.td:1315]: Maybe PseudoLUIADDI?  I agree that LIaddr isn't super clear.

You should also add a comment above the pseudo which describes its purpose.

**Review by preames — COMMENTED** [review]
General idea makes sense, and I see nothing concerning in the diffs.

**Pull request closed without merging**